### ISSUE TITLE
Exclusive Join not working when embedded in Fork/Join

### ISSUE THREAD

**Issue by RemcoBuddelmeijer** [Netflix/conductor#1240]
# Exclusive Join not working when embedded in Fork/Join
## Problem
Hello,

In one of our workflows we have a very specific with the following (generic) flow:
*Note: the task types assigned to any TASK step(s) are not bound to any specific task type, this can be either a JOIN/FORK, LAMBDA, SIMPLE, etc.. task type.*

```
FORK(T1) -> |
             TASK*(T2)                                          -> |
             DECISION(T2) -> TASK*(T3|T4) -> EXCLUSIVE JOIN(T5) -> |
                                                                    JOIN(T6)
```
When we embed a decision with an exclusive join inside another fork/join, the workflow will not continue processing after the exclusive join and with that never completing the parent join.
This only happens when we have the parent join (T6), join on the exclusive join (T5) (which is valid, see "Working alternative").

![image](https://user-images.githubusercontent.com/14148996/62627416-a8e6b380-b929-11e9-9c9d-66862a79605a.png)

Failing decision metadata representation:
```json
{
  "name" : "exclusive_join_test",
  "description" : "Exclusive Join Test",
  "version" : 1,
  "tasks" : [
    {
      "taskReferenceName" : "fork",
      "name" : "fork",
      "forkTasks" : [
        [
          {
            "name" : "lambda_nothing",
            "taskReferenceName" : "lambda_nothing",
            "inputParameters" : {
              "scriptExpression" : "return {value: \"nothing\"}"
            },
            "type" : "LAMBDA",
            "optional" : false,
            "asyncComplete" : false
          }
        ],
        [
          {
            "name" : "decision_test",
            "taskReferenceName" : "decision_test",
            "inputParameters" : {
              "case_value_param" : "${workflow.input.decision}"
            },
            "type" : "DECISION",
            "caseValueParam" : "case_value_param",
            "decisionCases" : {
              "true" : [
                {
                  "taskReferenceName" : "lambda_true",
                  "name" : "lambda_true",
                  "inputParameters" : {
                    "value" : "${workflow.input.decision}",
                    "scriptExpression" : "return {value: $.value}"
                  },
                  "optional" : false,
                  "type" : "LAMBDA",
                  "asyncComplete" : false
                }
              ]
            },
            "defaultCase" : [
              {
                "name" : "lambda_false",
                "taskReferenceName" : "lambda_false",
                "inputParameters" : {
                  "value" : "${workflow.input.decision}",
                  "scriptExpression" : "return {value: $.value}"
                },
                "type" : "LAMBDA",
                "optional" : false,
                "asyncComplete" : false
              }
            ],
            "optional" : false,
            "asyncComplete" : false
          },
          {
            "name" : "exclusive_join",
            "taskReferenceName" : "exclusive_join",
            "type" : "EXCLUSIVE_JOIN",
            "joinOn" : [
              "lambda_true",
              "lambda_false"
            ],
            "optional" : false,
            "defaultExclusiveJoinTask" : [
              "decision_test"
            ],
            "asyncComplete" : false
          }
        ]
      ],
      "type" : "FORK_JOIN"
    },
    {
      "name" : "join",
      "taskReferenceName" : "join",
      "joinOn" : [
        "lambda_nothing",
        "exclusive_join"
      ],
      "type" : "JOIN"
    }
  ],
  "inputParameters" : [
    "decision"
  ],
  "schemaVersion" : 2,
  "restartable" : true,
  "workflowStatusListenerEnabled" : false
}
```

## Working alternative

However, when we isolate the same exclusive join outside of the fork/join, the exclusive join completes and the workflow continues as expected:

![image](https://user-images.githubusercontent.com/14148996/62630163-b3f01280-b92e-11e9-855f-89a856cb46be.png)

Working decision metadata representation:
```json
{
  "name" : "exclusive_join_test",
  "description" : "Exclusive Join Test",
  "version" : 1,
  "tasks" : [
    {
      "name" : "decision_test",
      "taskReferenceName" : "decision_test",
      "inputParameters" : {
        "case_value_param" : "${workflow.input.decision}"
      },
      "type" : "DECISION",
      "caseValueParam" : "case_value_param",
      "decisionCases" : {
        "true" : [
          {
            "taskReferenceName" : "lambda_true",
            "name" : "lambda_true",
            "inputParameters" : {
              "value" : "${workflow.input.decision}",
              "scriptExpression" : "return {value: $.value}"
            },
            "optional" : false,
            "type" : "LAMBDA",
            "asyncComplete" : false
          }
        ]
      },
      "defaultCase" : [
        {
          "name" : "lambda_false",
          "taskReferenceName" : "lambda_false",
          "inputParameters" : {
            "value" : "${workflow.input.decision}",
            "scriptExpression" : "return {value: $.value}"
          },
          "type" : "LAMBDA",
          "optional" : false,
          "asyncComplete" : false
        }
      ],
      "optional" : false,
      "asyncComplete" : false
    },
    {
      "name" : "exclusive_join",
      "taskReferenceName" : "exclusive_join",
      "type" : "EXCLUSIVE_JOIN",
      "joinOn" : [
        "lambda_true",
        "lambda_false"
      ],
      "optional" : false,
      "defaultExclusiveJoinTask" : [
        "decision_test"
      ],
      "asyncComplete" : false
    }
  ],
  "inputParameters" : [
    "decision"
  ],
  "schemaVersion" : 2,
  "restartable" : true,
  "workflowStatusListenerEnabled" : false
}
```

## Some metadata:
**Conductor version:** 2.10.2

**Comment by jkaipa** [comment]
Looks like this is linked/similar to https://github.com/Netflix/conductor/issues/644

**Comment by RemcoBuddelmeijer** [comment]
I looked at both the issues, I believe they are indeed connected in a certain way. 
After looking at both #644 and #881 I went to look for a workaround.
First, to validate the connection between both tickets, I changed my workflow to have yet another decision in the FORK/JOIN. 
Now the other task in the FORK is no longer continuing after the decision, so that bug is connected to here (see image 1).
However, I did find a way on how to have the same expected output, but in a different order (similar to "fix" in #881), for that: see image 2.
I hope that this is of any help.

![image](https://user-images.githubusercontent.com/14148996/62654683-9e93dc00-b960-11e9-93dc-25794eb3a7a3.png)

The content of the workflow definition:
```json
{
  "name" : "exclusive_join_test",
  "description" : "Exclusive Join Test",
  "version" : 1,
  "tasks" : [
    {
      "taskReferenceName" : "fork",
      "name" : "fork",
      "forkTasks" : [
        [
          {
            "name" : "lambda_fork",
            "taskReferenceName" : "lambda_fork",
            "forkTasks" : [
              [
                {
                  "taskReferenceName" : "lambda_something",
                  "name" : "lambda_something",
                  "inputParameters" : {
                    "scriptExpression" : "return {value: \"something\"}"
                  },
                  "optional" : false,
                  "type" : "LAMBDA",
                  "asyncComplete" : false
                }
              ],
              [
                {
                  "taskReferenceName" : "lambda_nothing",
                  "name" : "lambda_nothing",
                  "inputParameters" : {
                    "scriptExpression" : "return {value: \"nothing\"}"
                  },
                  "optional" : false,
                  "type" : "LAMBDA",
                  "asyncComplete" : false
                }
              ]
            ],
            "type" : "FORK_JOIN",
            "optional" : false,
            "asyncComplete" : false
          },
          {
            "name" : "lambda_join",
            "taskReferenceName" : "lambda_join",
            "joinOn" : [
              "lambda_nothing",
              "lambda_something"
            ],
            "type" : "JOIN"
          },
          {
            "name" : "decision_lambda",
            "taskReferenceName" : "decision_lambda",
            "inputParameters" : {
              "case_value_param" : "${lambda_join.output.lambda_something.result.value}"
            },
            "type" : "DECISION",
            "caseValueParam" : "case_value_param",
            "decisionCases" : {
              "something" : [
                {
                  "taskReferenceName" : "lambda_otherthing",
                  "name" : "lambda_otherthing",
                  "inputParameters" : {
                    "scriptExpression" : "return {value: \"otherthing\"}"
                  },
                  "optional" : false,
                  "type" : "LAMBDA",
                  "asyncComplete" : false
                }
              ]
            },
            "defaultCase" : [
              {
                "name" : "lambda_mything",
                "taskReferenceName" : "lambda_mything",
                "inputParameters" : {
                  "scriptExpression" : "return {value: \"mything\"}"
                },
                "type" : "LAMBDA",
                "optional" : false,
                "asyncComplete" : false
              }
            ],
            "optional" : false,
            "asyncComplete" : false
          }
        ],
        [
          {
            "name" : "decision_test",
            "taskReferenceName" : "decision_test",
            "inputParameters" : {
              "case_value_param" : "${workflow.input.decision}"
            },
            "type" : "DECISION",
            "caseValueParam" : "case_value_param",
            "decisionCases" : {
              "true" : [
                {
                  "taskReferenceName" : "lambda_true",
                  "name" : "lambda_true",
                  "inputParameters" : {
                    "value" : "${workflow.input.decision}",
                    "scriptExpression" : "return {value: $.value}"
                  },
                  "optional" : false,
                  "type" : "LAMBDA",
                  "asyncComplete" : false
                }
              ]
            },
            "defaultCase" : [
              {
                "name" : "lambda_false",
                "taskReferenceName" : "lambda_false",
                "inputParameters" : {
                  "value" : "${workflow.input.decision}",
                  "scriptExpression" : "return {value: $.value}"
                },
                "type" : "LAMBDA",
                "optional" : false,
                "asyncComplete" : false
              }
            ],
            "optional" : false,
            "asyncComplete" : false
          }
        ]
      ],
      "type" : "FORK_JOIN"
    },
    {
      "name" : "join",
      "taskReferenceName" : "join",
      "joinOn" : [
        "decision_lambda"
      ],
      "type" : "JOIN"
    },
    {
      "name" : "exclusive_join",
      "taskReferenceName" : "exclusive_join",
      "type" : "EXCLUSIVE_JOIN",
      "joinOn" : [
        "lambda_true",
        "lambda_false"
      ],
      "optional" : false,
      "asyncComplete" : false
    }
  ],
  "inputParameters" : [
    "decision"
  ],
  "schemaVersion" : 2,
  "restartable" : true,
  "workflowStatusListenerEnabled" : false
}
```

**Comment by apanicker-nflx** [comment]
Thanks for reporting the issue. We will look into fixing this issue soon. However, feel free to submit a PR with the fix if you need this sooner.

**Comment by jkaipa** [comment]
@apanicker-nflx This is a duplicate of https://github.com/Netflix/conductor/issues/644 .

**Comment by apanicker-nflx** [comment]
Fixed as part of #1449
